File: modules/core/src/data/java/galena/oreganized/data/extensions/OBlockStateExtensions.java
package galena.oreganized.data.extensions;

import static net.minecraft.resources.ResourceLocation.withDefaultNamespace;
import static net.neoforged.neoforge.client.model.generators.ModelProvider.BLOCK_FOLDER;
import static net.neoforged.neoforge.client.model.generators.ModelProvider.ITEM_FOLDER;

import net.minecraft.core.Direction;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.level.block.*;
import net.minecraft.world.level.block.state.properties.BlockStateProperties;
import net.neoforged.neoforge.client.model.generators.*;
import net.neoforged.neoforge.registries.DeferredBlock;

public class OBlockStateExtensions {

    public static final ResourceLocation CUTOUT = withDefaultNamespace("cutout");
    public static final ResourceLocation TRANSLUCENT = withDefaultNamespace("translucent");

    public static void cubeAll(BlockStateProvider provider, DeferredBlock<? extends Block> block) {
        var model = provider.cubeAll(block.value());
        provider.simpleBlockWithItem(block.value(), model);
    }

    public static void slab(BlockStateProvider provider, DeferredBlock<? extends Block> block, DeferredBlock<? extends SlabBlock> slab) {
        var texture = provider.blockTexture(block.value());
        provider.slabBlock(slab.value(), texture, texture);
        blockItem(provider, slab);
    }

    public static void stairs(BlockStateProvider provider, DeferredBlock<? extends Block> block, DeferredBlock<? extends StairBlock> stairs) {
        var texture = provider.blockTexture(block.value());
        provider.stairsBlock(stairs.value(), texture);
        blockItem(provider, stairs);
    }

    public static void wall(BlockStateProvider provider, DeferredBlock<? extends Block> block, DeferredBlock<? extends WallBlock> wall) {
        var texture = provider.blockTexture(block.value());
        provider.wallBlock(wall.value(), texture);
        provider.itemModels()
                .getBuilder(wall.getId().getPath())
                .parent(provider.models().wallInventory(wall.getId().getPath() + "_inventory", texture));
    }

    public static void waxed(BlockStateProvider provider, DeferredBlock<? extends Block> block, Block origin) {
        var parent = provider.blockTexture(origin);
        var model = provider.models().getExistingFile(parent);
        provider.simpleBlockWithItem(block.value(), model);
    }

    public static void trapDoor(BlockStateProvider provider, DeferredBlock<? extends TrapDoorBlock> block) {
        var texture = provider.blockTexture(block.value());
        provider.trapdoorBlock(block.value(), texture, true);
        provider.simpleBlockItem(block.value(), provider.models().getExistingFile(texture.withSuffix("_bottom")));
    }

    public static void door(BlockStateProvider provider, DeferredBlock<? extends DoorBlock> block) {
        var texture = provider.blockTexture(block.value());
        provider.doorBlock(block.value(), texture.withSuffix("_bottom"), texture.withSuffix("_top"));
        generatedItem(provider, block, ITEM_FOLDER);
    }

    /**
     * copied from {@link com.teamabnormals.blueprint.core.data.client.BlueprintBlockStateProvider}
     */
    private static void pane(BlockStateProvider provider, Block block, ModelFile post, ModelFile postEnds, ModelFile side, ModelFile sideAlt, ModelFile cap, ModelFile capAlt) {
        MultiPartBlockStateBuilder builder = provider.getMultipartBuilder(block).part().modelFile(postEnds).addModel().end();
        builder.part().modelFile(post).addModel().condition(BlockStateProperties.NORTH, false).condition(BlockStateProperties.WEST, false).condition(BlockStateProperties.SOUTH, false).condition(BlockStateProperties.EAST, false).end();

        for (Direction direction : Direction.Plane.HORIZONTAL.stream().toList()) {
            builder.part().modelFile(direction == Direction.SOUTH || direction == Direction.WEST ? capAlt : cap).rotationY(direction.getAxis() == Direction.Axis.X ? 90 : 0).addModel()
                    .condition(BlockStateProperties.NORTH, PipeBlock.PROPERTY_BY_DIRECTION.get(direction) == BlockStateProperties.NORTH)
                    .condition(BlockStateProperties.WEST, PipeBlock.PROPERTY_BY_DIRECTION.get(direction) == BlockStateProperties.WEST)
                    .condition(BlockStateProperties.SOUTH, PipeBlock.PROPERTY_BY_DIRECTION.get(direction) == BlockStateProperties.SOUTH)
                    .condition(BlockStateProperties.EAST, PipeBlock.PROPERTY_BY_DIRECTION.get(direction) == BlockStateProperties.EAST)
                    .end();

        }

        PipeBlock.PROPERTY_BY_DIRECTION.forEach((dir, value) -> {
            if (dir.getAxis().isHorizontal()) {
                builder.part().modelFile(dir == Direction.SOUTH || dir == Direction.WEST ? sideAlt : side).rotationY(dir.getAxis() == Direction.Axis.X ? 90 : 0).addModel().condition(value, true).end();
            }
        });
    }

    /**
     * copied from {@link com.teamabnormals.blueprint.core.data.client.BlueprintBlockStateProvider}
     */
    public static void bars(BlockStateProvider provider, DeferredBlock<? extends IronBarsBlock> block) {
        var texture = provider.blockTexture(block.value());
        var name = block.getId().getPath();
        var edgeTexture = texture.withSuffix("_edge");

        var post = ironBarsModel(provider, name, "post", texture).texture("bars", edgeTexture);
        var postEnds = ironBarsModel(provider, name, "post_ends", texture).texture("edge", edgeTexture);
        var side = ironBarsModel(provider, name, "side", texture).texture("bars", texture).texture("edge", edgeTexture);
        var sideAlt = ironBarsModel(provider, name, "side_alt", texture).texture("bars", texture).texture("edge", edgeTexture);
        var cap = ironBarsModel(provider, name, "cap", texture).texture("bars", texture).texture("edge", edgeTexture);
        var capAlt = ironBarsModel(provider, name, "cap_alt", texture).texture("bars", texture).texture("edge", edgeTexture);

        pane(provider, block.value(), post, postEnds, side, sideAlt, cap, capAlt);
        generatedItem(provider, block, BLOCK_FOLDER);
    }

    public static BlockModelBuilder ironBarsModel(BlockStateProvider provider, String name, String suffix, ResourceLocation texture) {
        return provider.models()
                .withExistingParent(name + "_" + suffix, withDefaultNamespace("block/iron_bars_" + suffix))
                .renderType(CUTOUT)
                .texture("particle", texture);
    }

    public static void pillar(BlockStateProvider provider, DeferredBlock<? extends RotatedPillarBlock> block) {
        provider.logBlock(block.value());
        blockItem(provider, block);
    }

    public static void crossWithPot(BlockStateProvider provider, DeferredBlock<? extends Block> cross, DeferredBlock<? extends FlowerPotBlock> potted) {
        var texture = provider.blockTexture(cross.value());
        crossBlock(provider, cross);
        var pottedModel = provider.models().singleTexture(potted.getId().getPath(), withDefaultNamespace("block/flower_pot_cross"), "plant", texture).renderType(CUTOUT);
        provider.simpleBlock(potted.value(), pottedModel);
    }

    public static void crossBlock(BlockStateProvider provider, DeferredBlock<? extends Block> cross) {
        var texture = provider.blockTexture(cross.value());
        provider.simpleBlock(cross.value(), provider.models().cross(cross.getId().getPath(), texture).renderType(CUTOUT));
        generatedItem(provider, cross, BLOCK_FOLDER);
    }

    public static ItemModelBuilder generatedItem(BlockStateProvider provider, DeferredBlock<? extends Block> block, String type) {
        return OItemModelExtensions.generatedItem(provider.itemModels(), block.getId(), block.getId(), type);
    }

    public static ModelFile cauldronModel(BlockStateProvider provider, DeferredBlock<? extends Block> block, ResourceLocation texture, int age) {
        var name = block.getId().getPath() + "_" + age;
        return provider.models().withExistingParent(name, BLOCK_FOLDER + "/template_cauldron_full")
                .texture("content", texture);
    }

    public static void cubeBottomTopBlock(BlockStateProvider provider, DeferredBlock<? extends Block> block) {
        var texture = provider.blockTexture(block.value());
        cubeBottomTopBlock(provider, block, texture.withSuffix("_side"), texture.withSuffix("_bottom"), texture.withSuffix("_top"));
    }

    public static void cubeBottomTopBlock(BlockStateProvider provider, DeferredBlock<? extends Block> block, ResourceLocation sideTexture, ResourceLocation bottomTexture, ResourceLocation topTexture) {
        provider.simpleBlock(block.value(), provider.models().cubeBottomTop(block.getId().getPath(), sideTexture, bottomTexture, topTexture));
        blockItem(provider, block);
    }

    public static void blockItem(BlockStateProvider provider, DeferredBlock<? extends Block> block) {
        blockItem(provider, block, provider.blockTexture(block.value()));
    }

    public static void blockItem(BlockStateProvider provider, DeferredBlock<? extends Block> block, ResourceLocation model) {
        var existing = provider.models().getExistingFile(model);
        provider.simpleBlockItem(block.value(), existing);
    }

    public static ResourceLocation blockTexture(ResourceLocation id) {
        return id.withPrefix(BLOCK_FOLDER + "/");
    }

}

File: modules/core/src/data/kotlin/galena/oreganized/data/extensions/OBlockStateExtensions.kt
@file:JvmName("OBlockStateExtensions")

package galena.oreganized.data.extensions

import galena.oreganized.index.sets.StoneSet
import net.minecraft.core.Direction
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.level.block.*
import net.minecraft.world.level.block.state.properties.BlockStateProperties
import net.minecraft.world.level.block.state.properties.Property
import net.neoforged.neoforge.client.model.generators.*
import net.neoforged.neoforge.registries.DeferredHolder
import java.util.function.BiConsumer

@JvmField
val CUTOUT: ResourceLocation = ResourceLocation.withDefaultNamespace("cutout")

@JvmField
val TRANSLUCENT: ResourceLocation = ResourceLocation.withDefaultNamespace("translucent")

fun BlockStateProvider.cubeAll(block: DeferredHolder<Block, out Block>) {
    val model = cubeAll(block.value())
    simpleBlockWithItem(block.value(), model)
}

fun BlockStateProvider.slab(
    block: DeferredHolder<Block, out Block>,
    slab: DeferredHolder<Block, out SlabBlock>,
) {
    val texture = blockTexture(block.value())
    slabBlock(slab.value(), texture, texture)
    blockItem(slab)
}

fun BlockStateProvider.stairs(
    block: DeferredHolder<Block, out Block>,
    stairs: DeferredHolder<Block, out StairBlock>,
) {
    val texture = blockTexture(block.value())
    stairsBlock(stairs.value(), texture)
    blockItem(stairs)
}

fun BlockStateProvider.wall(
    block: DeferredHolder<Block, out Block>,
    wall: DeferredHolder<Block, out WallBlock>,
) {
    val texture = blockTexture(block.value())
    wallBlock(wall.value(), texture)
    itemModels()
        .getBuilder(wall.getId().getPath())
        .parent(models().wallInventory(wall.getId().getPath() + "_inventory", texture))
}

fun BlockStateProvider.waxed(
    block: DeferredHolder<Block, out Block>,
    origin: Block,
) {
    val parent = blockTexture(origin)
    val model = models().getExistingFile(parent)
    simpleBlockWithItem(block.value(), model)
}

fun BlockStateProvider.trapDoor(block: DeferredHolder<Block, out TrapDoorBlock>) {
    val texture = blockTexture(block.value())
    trapdoorBlock(block.value(), texture, true)
    simpleBlockItem(block.value(), models().getExistingFile(texture.withSuffix("_bottom")))
}

fun BlockStateProvider.door(block: DeferredHolder<Block, out DoorBlock>) {
    val texture = blockTexture(block.value())
    doorBlock(block.value(), texture.withSuffix("_bottom"), texture.withSuffix("_top"))
    generatedItem(block, ModelProvider.ITEM_FOLDER)
}

/**
 * copied from [com.teamabnormals.blueprint.core.data.client.BlueprintBlockStateProvider]
 */
fun BlockStateProvider.pane(
    block: Block,
    post: ModelFile,
    postEnds: ModelFile,
    side: ModelFile,
    sideAlt: ModelFile,
    cap: ModelFile,
    capAlt: ModelFile,
) {
    val builder =
        getMultipartBuilder(block)
            .part()
            .modelFile(postEnds)
            .addModel()
            .end()
    builder
        .part()
        .modelFile(post)
        .addModel()
        .condition(BlockStateProperties.NORTH, false)
        .condition(BlockStateProperties.WEST, false)
        .condition(BlockStateProperties.SOUTH, false)
        .condition(BlockStateProperties.EAST, false)
        .end()

    for (direction in Direction.Plane.HORIZONTAL
        .stream()
        .toList()) {
        builder
            .part()
            .modelFile(if (direction == Direction.SOUTH || direction == Direction.WEST) capAlt else cap)
            .rotationY(if (direction.getAxis() === Direction.Axis.X) 90 else 0)
            .addModel()
            .condition(
                BlockStateProperties.NORTH,
                PipeBlock.PROPERTY_BY_DIRECTION.get(direction) === BlockStateProperties.NORTH,
            ).condition(
                BlockStateProperties.WEST,
                PipeBlock.PROPERTY_BY_DIRECTION.get(direction) === BlockStateProperties.WEST,
            ).condition(
                BlockStateProperties.SOUTH,
                PipeBlock.PROPERTY_BY_DIRECTION.get(direction) === BlockStateProperties.SOUTH,
            ).condition(
                BlockStateProperties.EAST,
                PipeBlock.PROPERTY_BY_DIRECTION.get(direction) === BlockStateProperties.EAST,
            ).end()
    }

    directions { dir, prop ->
        if (dir.getAxis().isHorizontal()) {
            builder
                .part()
                .modelFile(if (dir == Direction.SOUTH || dir == Direction.WEST) sideAlt else side)
                .rotationY(if (dir.getAxis() === Direction.Axis.X) 90 else 0)
                .addModel()
                .condition(prop, true)
                .end()
        }
    }
}

/**
 * copied from [com.teamabnormals.blueprint.core.data.client.BlueprintBlockStateProvider]
 */
fun BlockStateProvider.bars(block: DeferredHolder<Block, out IronBarsBlock>) {
    val texture = blockTexture(block.value())
    val name = block.getId().getPath()
    val edgeTexture = texture.withSuffix("_edge")

    val post = ironBarsModel(name, "post", texture).texture("bars", edgeTexture)
    val postEnds = ironBarsModel(name, "post_ends", texture).texture("edge", edgeTexture)
    val side = ironBarsModel(name, "side", texture).texture("bars", texture).texture("edge", edgeTexture)
    val sideAlt =
        ironBarsModel(name, "side_alt", texture).texture("bars", texture).texture("edge", edgeTexture)
    val cap = ironBarsModel(name, "cap", texture).texture("bars", texture).texture("edge", edgeTexture)
    val capAlt =
        ironBarsModel(name, "cap_alt", texture).texture("bars", texture).texture("edge", edgeTexture)

    pane(block.value(), post, postEnds, side, sideAlt, cap, capAlt)
    generatedItem(block, ModelProvider.BLOCK_FOLDER)
}

fun BlockStateProvider.ironBarsModel(
    name: String,
    suffix: String,
    texture: ResourceLocation,
): BlockModelBuilder =
    models()
        .withExistingParent(name + "_" + suffix, ResourceLocation.withDefaultNamespace("block/iron_bars_" + suffix))
        .renderType(CUTOUT)
        .texture("particle", texture)

fun BlockStateProvider.pillar(block: DeferredHolder<Block, out RotatedPillarBlock>) {
    logBlock(block.value())
    blockItem(block)
}

fun BlockStateProvider.cubeColumn(block: DeferredHolder<Block, out Block>) {
    val texture = blockTexture(block.value())
    cubeBottomTopBlock(
        block,
        texture.withSuffix("_side"),
        texture.withSuffix("_top"),
        texture.withSuffix("_top"),
    )
}

fun BlockStateProvider.crossWithPot(
    cross: DeferredHolder<Block, out Block>,
    potted: DeferredHolder<Block, out FlowerPotBlock>,
) {
    val texture = blockTexture(cross.value())
    crossBlock(cross)
    val pottedModel =
        models()
            .singleTexture(
                potted.getId().getPath(),
                ResourceLocation.withDefaultNamespace("block/flower_pot_cross"),
                "plant",
                texture,
            ).renderType(
                CUTOUT,
            )
    simpleBlock(potted.value(), pottedModel)
}

fun BlockStateProvider.crossBlock(cross: DeferredHolder<Block, out Block>) {
    val texture = blockTexture(cross.value())
    simpleBlock(
        cross.value(),
        models().cross(cross.getId().getPath(), texture).renderType(CUTOUT),
    )
    generatedItem(cross, ModelProvider.BLOCK_FOLDER)
}

fun BlockStateProvider.generatedItem(
    block: DeferredHolder<Block, out Block>,
    type: String,
): ItemModelBuilder = itemModels().generatedItem(block.getId(), block.getId(), type)

fun BlockStateProvider.cauldronModel(
    block: DeferredHolder<Block, out Block>,
    texture: ResourceLocation,
    age: Int,
): ModelFile {
    val name = block.getId().getPath() + "_" + age
    return cauldronModel(name, texture)
}

fun BlockStateProvider.cauldronModel(
    name: String,
    texture: ResourceLocation,
): ModelFile =
    models()
        .withExistingParent(name, ModelProvider.BLOCK_FOLDER + "/template_cauldron_full")
        .texture("content", texture)

fun BlockStateProvider.cubeBottomTopBlock(block: DeferredHolder<Block, out Block>) {
    val texture = blockTexture(block.value())
    cubeBottomTopBlock(
        block,
        texture.withSuffix("_side"),
        texture.withSuffix("_bottom"),
        texture.withSuffix("_top"),
    )
}

fun BlockStateProvider.cubeBottomTopBlock(
    block: DeferredHolder<Block, out Block>,
    sideTexture: ResourceLocation,
    bottomTexture: ResourceLocation,
    topTexture: ResourceLocation,
) {
    simpleBlock(
        block.value(),
        models().cubeBottomTop(block.getId().getPath(), sideTexture, bottomTexture, topTexture),
    )
    blockItem(block)
}

@JvmOverloads
fun BlockStateProvider.blockItem(
    block: DeferredHolder<Block, out Block>,
    model: ResourceLocation = blockTexture(block.value()),
) {
    val existing = models().getExistingFile(model)
    simpleBlockItem(block.value(), existing)
}

fun ResourceLocation.blockTexture(): ResourceLocation = withPrefix(ModelProvider.BLOCK_FOLDER + "/")

fun BlockStateProvider.stoneSet(block: StoneSet<*, *, *, *>) {
    cubeAll(block.block)
    stairs(block.block, block.stairs)
    slab(block.block, block.slab)
    wall(block.block, block.wall)
}

fun horizontalDirections(consumer: BiConsumer<Direction, Property<Boolean>>) {
    CrossCollisionBlock.PROPERTY_BY_DIRECTION
        .asSequence()
        .sortedBy { it.key.ordinal }
        .forEach { entry ->
            consumer.accept(
                entry.key,
                entry.value,
            )
        }
}

fun directions(consumer: BiConsumer<Direction, Property<Boolean>>) {
    PipeBlock.PROPERTY_BY_DIRECTION
        .asSequence()
        .sortedBy { it.key.ordinal }
        .forEach { entry ->
            consumer.accept(
                entry.key,
                entry.value,
            )
        }
}
